File: aruco_imu_eskf_localization/aruco_imu_eskf_localization/common/camera_calibration.py
from __future__ import annotations

from dataclasses import dataclass

import cv2
import numpy as np
import yaml


@dataclass(frozen=True)
class CameraCalibration:
    camera_model: str
    camera_matrix: np.ndarray
    distortion_coefficients: np.ndarray
    image_width: int
    image_height: int
    used_legacy_default: bool = False

    @property
    def image_size(self) -> tuple[int, int]:
        return self.image_width, self.image_height


def load_camera_calibration(path: str) -> CameraCalibration:
    with open(path, 'r', encoding='utf-8') as file_obj:
        calib = yaml.safe_load(file_obj)

    raw_camera_model = str(calib.get('camera_model', '')).strip().lower()
    used_legacy_default = False
    if not raw_camera_model:
        raw_camera_model = 'fisheye'
        used_legacy_default = True

    if raw_camera_model != 'fisheye':
        raise ValueError(f'unsupported camera_model {raw_camera_model!r}; expected fisheye')

    camera_matrix = np.asarray(
        calib['camera_matrix']['data'],
        dtype=np.float64,
    ).reshape(3, 3)
    distortion_coefficients = np.asarray(
        calib['distortion_coefficients']['data'],
        dtype=np.float64,
    ).reshape(-1, 1)
    if distortion_coefficients.shape[0] != 4:
        raise ValueError(
            'fisheye calibration requires exactly 4 distortion coefficients'
        )

    image_size = calib.get('image_size') or {}
    image_width = int(image_size['width'])
    image_height = int(image_size['height'])

    return CameraCalibration(
        camera_model=raw_camera_model,
        camera_matrix=camera_matrix,
        distortion_coefficients=distortion_coefficients,
        image_width=image_width,
        image_height=image_height,
        used_legacy_default=used_legacy_default,
    )


def build_fisheye_rectification(
    calibration: CameraCalibration,
    balance: float = 0.0,
) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    if calibration.camera_model != 'fisheye':
        raise ValueError(
            f'unsupported camera_model {calibration.camera_model!r}; expected fisheye'
        )

    image_size = calibration.image_size
    rectified_camera_matrix = cv2.fisheye.estimateNewCameraMatrixForUndistortRectify(
        calibration.camera_matrix,
        calibration.distortion_coefficients,
        image_size,
        np.eye(3, dtype=np.float64),
        balance=float(np.clip(balance, 0.0, 1.0)),
    )
    map1, map2 = cv2.fisheye.initUndistortRectifyMap(
        calibration.camera_matrix,
        calibration.distortion_coefficients,
        np.eye(3, dtype=np.float64),
        rectified_camera_matrix,
        image_size,
        cv2.CV_32FC1,
    )
    zero_distortion = np.zeros((4, 1), dtype=np.float64)
    return rectified_camera_matrix, zero_distortion, map1, map2

File: aruco_imu_eskf_localization_cpp/src/camera_calibration.cpp
#include "aruco_imu_eskf_localization_cpp/camera_calibration.hpp"

#include <opencv2/calib3d.hpp>
#include <yaml-cpp/yaml.h>

#include <algorithm>
#include <stdexcept>
#include <string>

namespace aruco_imu_eskf_localization_cpp
{

namespace
{

void collect_numeric_data(const YAML::Node & node, std::vector<double> & out)
{
  if (!node.IsSequence()) {
    out.push_back(node.as<double>());
    return;
  }
  for (const auto & child : node) {
    collect_numeric_data(child, out);
  }
}

std::vector<double> numeric_data(const YAML::Node & node)
{
  std::vector<double> out;
  collect_numeric_data(node, out);
  return out;
}

}  // namespace

CameraCalibration load_camera_calibration(const std::string & path)
{
  const YAML::Node root = YAML::LoadFile(path);
  CameraCalibration calibration;
  calibration.camera_model = root["camera_model"] ? root["camera_model"].as<std::string>() : "fisheye";
  calibration.used_legacy_default = !root["camera_model"];
  if (calibration.camera_model != "fisheye") {
    throw std::runtime_error("unsupported camera_model " + calibration.camera_model);
  }

  const auto camera_data = numeric_data(root["camera_matrix"]["data"]);
  if (camera_data.size() != 9) {
    throw std::runtime_error("camera_matrix.data must contain 9 values");
  }
  calibration.camera_matrix = cv::Mat(3, 3, CV_64F);
  for (int i = 0; i < 9; ++i) {
    calibration.camera_matrix.at<double>(i / 3, i % 3) = camera_data.at(i);
  }

  const auto dist_data = numeric_data(root["distortion_coefficients"]["data"]);
  if (dist_data.size() != 4) {
    throw std::runtime_error("fisheye distortion_coefficients.data must contain 4 values");
  }
  calibration.distortion_coefficients = cv::Mat(4, 1, CV_64F);
  for (int i = 0; i < 4; ++i) {
    calibration.distortion_coefficients.at<double>(i, 0) = dist_data.at(i);
  }

  calibration.image_size = cv::Size(
    root["image_size"]["width"].as<int>(),
    root["image_size"]["height"].as<int>());
  return calibration;
}

FisheyeRectification build_fisheye_rectification(
  const CameraCalibration & calibration,
  double balance)
{
  FisheyeRectification rect;
  balance = std::clamp(balance, 0.0, 1.0);
  cv::fisheye::estimateNewCameraMatrixForUndistortRectify(
    calibration.camera_matrix,
    calibration.distortion_coefficients,
    calibration.image_size,
    cv::Matx33d::eye(),
    rect.rectified_camera_matrix,
    balance);
  cv::fisheye::initUndistortRectifyMap(
    calibration.camera_matrix,
    calibration.distortion_coefficients,
    cv::Matx33d::eye(),
    rect.rectified_camera_matrix,
    calibration.image_size,
    CV_32FC1,
    rect.map1,
    rect.map2);
  rect.zero_distortion = cv::Mat::zeros(4, 1, CV_64F);
  return rect;
}

}  // namespace aruco_imu_eskf_localization_cpp
